File: iosApp/LoaPalette/LoaPalette/Views/CardSearch/CardSearchViewModel.swift
//
//  CardSearchViewModel.swift
//  LoaPalette
//
//  Created by 片岡寿哉 on 2025/11/28.
//

import Combine
import SwiftUI

// 一時的な回避策: Kotlin側のクラスが認識されるまで、Swift側で完全に実装
// 参考: https://api-lorcana.com/#/Cards/get%20cards
@MainActor
class CardSearchViewModel: ObservableObject {
    @Published var cards: [LorcanaCard] = []
    @Published var searchState: SearchState = .idle
    @Published var errorMessage: String? = nil
    @Published var searchQuery: String = ""
    @Published var isLoadingMore: Bool = false
    @Published var hasMore: Bool = true

    private var searchTask: Task<Void, Never>?
    private var currentPage = 1
    private var currentSearchQuery: String? = nil
    private let pageSize = 20

    func search(query: String) {
        searchQuery = query
        searchTask?.cancel()

        // 新しい検索の場合はリセット
        if currentSearchQuery != query {
            currentPage = 1
            currentSearchQuery = query
            hasMore = true
        }

        if query.isEmpty {
            cards = []
            searchState = .idle
            return
        }

        searchState = .loading
        errorMessage = nil

        searchTask = Task { @MainActor in
            // デバウンス処理（500ms待機）
            try? await Task.sleep(nanoseconds: 500_000_000)

            await performSearch(query: query, page: currentPage)
        }
    }

    func loadAllCards() {
        // 新しい検索の場合はリセット
        currentPage = 1
        currentSearchQuery = nil
        hasMore = true

        searchState = .loading
        errorMessage = nil

        searchTask = Task { @MainActor in
            await performLoadAllCards(page: currentPage)
        }
    }

    func loadMore() {
        if isLoadingMore || !hasMore {
            return
        }

        currentPage += 1
        isLoadingMore = true

        Task { @MainActor in
            let query = currentSearchQuery ?? ""
            if query.isEmpty {
                await performLoadAllCards(page: currentPage)
            } else {
                await performSearch(query: query, page: currentPage)
            }
        }
    }

    func clear() {
        searchTask?.cancel()
        cards = []
        searchQuery = ""
        searchState = .idle
        errorMessage = nil
        currentPage = 1
        currentSearchQuery = nil
        hasMore = true
        isLoadingMore = false
    }

    func cleanup() {
        searchTask?.cancel()
        searchTask = nil
    }

    deinit {
        searchTask?.cancel()
        searchTask = nil
    }

    // API呼び出し
    // API仕様: https://api-lorcana.com/#/Cards/get%20cards
    private func performSearch(query: String, page: Int) async {
        // api-lorcana.comのエンドポイントを使用
        guard var urlComponents = URLComponents(string: "https://api-lorcana.com/cards")
        else {
            errorMessage = "無効なURLです"
            searchState = .error
            if page == 1 {
                cards = []
            } else {
                isLoadingMore = false
                currentPage -= 1
            }
            return
        }

        // 検索パラメータを追加
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "limit", value: "\(pageSize)"),
            URLQueryItem(name: "page", value: "\(page)"),
        ]

        // 検索クエリが空でない場合は追加
        if !query.isEmpty {
            queryItems.append(URLQueryItem(name: "search", value: query))
        }

        urlComponents.queryItems = queryItems

        guard let url = urlComponents.url else {
            errorMessage = "無効なURLです"
            searchState = .error
            if page == 1 {
                cards = []
            } else {
                isLoadingMore = false
                currentPage -= 1
            }
            return
        }

        do {
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.setValue("UTF-8", forHTTPHeaderField: "Accept-Charset")

            let (data, response) = try await URLSession.shared.data(for: request)

            // HTTPステータスコードを確認
            if let httpResponse = response as? HTTPURLResponse {
                guard (200...299).contains(httpResponse.statusCode) else {
                    errorMessage = "サーバーエラー: \(httpResponse.statusCode)"
                    searchState = .error
                    if page == 1 {
                        cards = []
                    } else {
                        isLoadingMore = false
                        currentPage -= 1
                    }
                    return
                }
            }

            let decoder = JSONDecoder()

            // エラーレスポンスかどうかを確認
            if let errorResponse = try? decoder.decode(APIError.self, from: data) {
                self.errorMessage = "APIエラー: \(errorResponse.details)"
                self.searchState = .error
                if page == 1 {
                    self.cards = []
                } else {
                    self.isLoadingMore = false
                    self.currentPage -= 1
                }
                return
            }

            // APIは直接カードの配列を返す
            let newCards = try decoder.decode([LorcanaCard].self, from: data)
            if page == 1 {
                self.cards = newCards
            } else {
                self.cards.append(contentsOf: newCards)
                self.isLoadingMore = false
            }
            self.hasMore = newCards.count >= pageSize
            self.searchState = .success
        } catch let decodingError as DecodingError {
            // デコードエラーの詳細を表示
            var errorDescription = "データのフォーマットが正しくありません"
            switch decodingError {
            case .typeMismatch(let type, let context):
                errorDescription = "型の不一致: \(type) - \(context.debugDescription)"
            case .valueNotFound(let type, let context):
                errorDescription = "値が見つかりません: \(type) - \(context.debugDescription)"
            case .keyNotFound(let key, let context):
                errorDescription = "キーが見つかりません: \(key.stringValue) - \(context.debugDescription)"
            case .dataCorrupted(let context):
                errorDescription = "データが破損しています: \(context.debugDescription)"
            @unknown default:
                errorDescription = "デコードエラー: \(decodingError.localizedDescription)"
            }
            self.errorMessage = errorDescription
            self.searchState = .error
            if page == 1 {
                self.cards = []
            } else {
                self.isLoadingMore = false
                self.currentPage -= 1
            }
        } catch {
            self.errorMessage = "エラー: \(error.localizedDescription)"
            self.searchState = .error
            if page == 1 {
                self.cards = []
            } else {
                self.isLoadingMore = false
                self.currentPage -= 1
            }
        }
    }

    private func performLoadAllCards(page: Int) async {
        // api-lorcana.comのエンドポイントを使用
        guard var urlComponents = URLComponents(string: "https://api-lorcana.com/cards") else {
            errorMessage = "無効なURLです"
            searchState = .error
            if page == 1 {
                cards = []
            } else {
                isLoadingMore = false
                currentPage -= 1
            }
            return
        }

        urlComponents.queryItems = [
            URLQueryItem(name: "limit", value: "\(pageSize)"),
            URLQueryItem(name: "page", value: "\(page)"),
        ]

        guard let url = urlComponents.url else {
            errorMessage = "無効なURLです"
            searchState = .error
            if page == 1 {
                cards = []
            } else {
                isLoadingMore = false
                currentPage -= 1
            }
            return
        }

        do {
            var request = URLRequest(url: url)
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.setValue("UTF-8", forHTTPHeaderField: "Accept-Charset")

            let (data, response) = try await URLSession.shared.data(for: request)

            // HTTPステータスコードを確認
            if let httpResponse = response as? HTTPURLResponse {
                guard (200...299).contains(httpResponse.statusCode) else {
                    errorMessage = "サーバーエラー: \(httpResponse.statusCode)"
                    searchState = .error
                    if page == 1 {
                        cards = []
                    } else {
                        isLoadingMore = false
                        currentPage -= 1
                    }
                    return
                }
            }

            let decoder = JSONDecoder()

            // エラーレスポンスかどうかを確認
            if let errorResponse = try? decoder.decode(APIError.self, from: data) {
                self.errorMessage = "APIエラー: \(errorResponse.details)"
                self.searchState = .error
                if page == 1 {
                    self.cards = []
                } else {
                    self.isLoadingMore = false
                    self.currentPage -= 1
                }
                return
            }

            // APIは直接カードの配列を返す
            let newCards = try decoder.decode([LorcanaCard].self, from: data)
            if page == 1 {
                self.cards = newCards
            } else {
                self.cards.append(contentsOf: newCards)
                self.isLoadingMore = false
            }
            self.hasMore = newCards.count >= pageSize
            self.searchState = .success
        } catch let decodingError as DecodingError {
            // デコードエラーの詳細を表示
            var errorDescription = "データのフォーマットが正しくありません"
            switch decodingError {
            case .typeMismatch(let type, let context):
                errorDescription = "型の不一致: \(type) - \(context.debugDescription)"
            case .valueNotFound(let type, let context):
                errorDescription = "値が見つかりません: \(type) - \(context.debugDescription)"
            case .keyNotFound(let key, let context):
                errorDescription = "キーが見つかりません: \(key.stringValue) - \(context.debugDescription)"
            case .dataCorrupted(let context):
                errorDescription = "データが破損しています: \(context.debugDescription)"
            @unknown default:
                errorDescription = "デコードエラー: \(decodingError.localizedDescription)"
            }
            self.errorMessage = errorDescription
            self.searchState = .error
            if page == 1 {
                self.cards = []
            } else {
                self.isLoadingMore = false
                self.currentPage -= 1
            }
        } catch {
            self.errorMessage = "エラー: \(error.localizedDescription)"
            self.searchState = .error
            if page == 1 {
                self.cards = []
            } else {
                self.isLoadingMore = false
                self.currentPage -= 1
            }
        }
    }
}

File: shared/src/commonMain/kotlin/com/loapalette/shared/CardSearchViewModel.kt
package com.loapalette.shared

import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

@OptIn(ExperimentalObjCName::class)
@ObjCName("SearchState", exact = true)
enum class SearchState {
    IDLE,
    LOADING,
    SUCCESS,
    ERROR
}

@OptIn(ExperimentalObjCName::class)
@ObjCName("CardSearchViewModel", exact = true)
class CardSearchViewModel {
    private val viewModelScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val apiClient = LorcanaApiClient(createHttpClient())

    private val _cards = MutableStateFlow<List<LorcanaCard>>(emptyList())
    val cards: StateFlow<List<LorcanaCard>> = _cards.asStateFlow()

    private val _searchState = MutableStateFlow(SearchState.IDLE)
    val searchState: StateFlow<SearchState> = _searchState.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _searchQuery = MutableStateFlow("")
    val searchQuery: StateFlow<String> = _searchQuery.asStateFlow()

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore.asStateFlow()

    private val _hasMore = MutableStateFlow(true)
    val hasMore: StateFlow<Boolean> = _hasMore.asStateFlow()

    private var searchJob: kotlinx.coroutines.Job? = null
    private var currentPage = 1
    private var currentSearchQuery: String? = null
    private val pageSize = 20

    fun search(query: String) {
        _searchQuery.value = query
        searchJob?.cancel()

        if (currentSearchQuery != query) {
            currentPage = 1
            currentSearchQuery = query
            _hasMore.value = true
        }

        if (query.isBlank()) {
            loadAllCards()
            return
        }

        searchJob =
                viewModelScope.launch {
                    _searchState.value = SearchState.LOADING
                    _errorMessage.value = null

                    delay(500)

                    val result =
                            apiClient.searchCards(
                                    searchQuery = query,
                                    page = currentPage,
                                    pageSize = pageSize
                            )
                    result.fold(
                            onSuccess = { response ->
                                if (currentPage == 1) {
                                    _cards.value = response.cards
                                } else {
                                    _cards.value = _cards.value + response.cards
                                }
                                _hasMore.value = response.cards.size >= pageSize
                                _searchState.value = SearchState.SUCCESS
                            },
                            onFailure = { error ->
                                _errorMessage.value = error.message ?: "検索に失敗しました"
                                _searchState.value = SearchState.ERROR
                                if (currentPage == 1) {
                                    _cards.value = emptyList()
                                }
                            }
                    )
                }
    }

    fun loadAllCards() {
        currentPage = 1
        currentSearchQuery = null
        _hasMore.value = true

        viewModelScope.launch {
            _searchState.value = SearchState.LOADING
            _errorMessage.value = null

            val result = apiClient.getAllCards(page = currentPage, pageSize = pageSize)
            result.fold(
                    onSuccess = { response ->
                        _cards.value = response.cards
                        _hasMore.value = response.cards.size >= pageSize
                        _searchState.value = SearchState.SUCCESS
                    },
                    onFailure = { error ->
                        _errorMessage.value = error.message ?: "カードの取得に失敗しました"
                        _searchState.value = SearchState.ERROR
                        _cards.value = emptyList()
                    }
            )
        }
    }

    fun loadMore() {
        if (_isLoadingMore.value || !_hasMore.value) {
            return
        }

        currentPage++
        _isLoadingMore.value = true

        viewModelScope.launch {
            val query = currentSearchQuery ?: ""
            val result =
                    if (query.isBlank()) {
                        apiClient.getAllCards(page = currentPage, pageSize = pageSize)
                    } else {
                        apiClient.searchCards(
                                searchQuery = query,
                                page = currentPage,
                                pageSize = pageSize
                        )
                    }

            result.fold(
                    onSuccess = { response ->
                        _cards.value = _cards.value + response.cards
                        _hasMore.value = response.cards.size >= pageSize
                        _isLoadingMore.value = false
                    },
                    onFailure = { error ->
                        currentPage--
                        _errorMessage.value = error.message ?: "追加読み込みに失敗しました"
                        _isLoadingMore.value = false
                    }
            )
        }
    }

    fun clear() {
        searchJob?.cancel()
        _cards.value = emptyList()
        _searchQuery.value = ""
        _searchState.value = SearchState.IDLE
        _errorMessage.value = null
        currentPage = 1
        currentSearchQuery = null
        _hasMore.value = true
        _isLoadingMore.value = false
    }

    fun getCards(): List<LorcanaCard> {
        return _cards.value
    }

    fun getSearchState(): String {
        return _searchState.value.name
    }

    fun getErrorMessage(): String? {
        return _errorMessage.value
    }

    fun getSearchQuery(): String {
        return _searchQuery.value
    }

    fun getIsLoadingMore(): Boolean {
        return _isLoadingMore.value
    }

    fun getHasMore(): Boolean {
        return _hasMore.value
    }

    fun cleanup() {
        searchJob?.cancel()
        searchJob = null
    }
}
